libazureinit/src/kvp.rs

```diff
@@ -0,0 +1,649 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module provides core functionality for handling telemetry tracing
+//! related to azure-init's interaction with Hyper-V KVP (Key-Value Pair) storage.
+//!
+//! # Constants
+//! - `HV_KVP_EXCHANGE_MAX_KEY_SIZE`: Defines the maximum key size for KVP exchange.
+//! - `HV_KVP_EXCHANGE_MAX_VALUE_SIZE`: Defines the maximum value size for KVP exchange.
+//! - `HV_KVP_AZURE_MAX_VALUE_SIZE`: Maximum value size before splitting into multiple slices.
+//!
+
+use std::{
+    fmt::{self as std_fmt, Write as std_write},
+    fs::{File, OpenOptions},
+    io::{self, Error, ErrorKind, Write},
+    os::unix::fs::MetadataExt,
+    path::Path,
+    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
+};
+
+use tracing::{
+    field::Visit,
+    span::{Attributes, Id},
+    Subscriber,
+};
+
+use tracing_subscriber::{
+    layer::Context as TracingContext, registry::LookupSpan, Layer,
+};
+
+use sysinfo::{System, SystemExt};
+
+use tokio::sync::{mpsc::UnboundedReceiver, mpsc::UnboundedSender};
+
+//use crate::tracing::{handle_event, handle_span};
+
+use chrono::{DateTime, Utc};
+use std::fmt;
+use uuid::Uuid;
+
+const HV_KVP_EXCHANGE_MAX_KEY_SIZE: usize = 512;
+const HV_KVP_EXCHANGE_MAX_VALUE_SIZE: usize = 2048;
+const HV_KVP_AZURE_MAX_VALUE_SIZE: usize = 1024;
+const EVENT_PREFIX: &str = concat!("azure-init-", env!("CARGO_PKG_VERSION"));
+
+/// A wrapper around `std::time::Instant` that provides convenient methods
+/// for time tracking in spans and events. Implements the `Deref` trait, allowing
+/// access to the underlying `Instant` methods.
+///
+/// This struct captures the start time of spans/events and measures the elapsed time.
+#[derive(Clone)]
+struct MyInstant(Instant);
+
+impl std::ops::Deref for MyInstant {
+    type Target = Instant;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl MyInstant {
+    pub fn now() -> Self {
+        MyInstant(Instant::now())
+    }
+}
+
+/// A custom visitor that captures the value of the `msg` field from a tracing event.
+/// It implements the `tracing::field::Visit` trait and records the value into
+/// a provided mutable string reference.
+///
+/// This visitor is primarily used in the `on_event` method of the `EmitKVPLayer`
+/// to extract event messages and log them as key-value pairs.
+pub struct StringVisitor<'a> {
+    string: &'a mut String,
+}
+
+impl Visit for StringVisitor<'_> {
+    /// Records the debug representation of the event's value and stores it in the provided string.
+    ///
+    /// # Arguments
+    /// * `_field` - A reference to the event's field metadata.
+    /// * `value` - The debug value associated with the field.
+    fn record_debug(
+        &mut self,
+        _field: &tracing::field::Field,
+        value: &dyn std_fmt::Debug,
+    ) {
+        write!(self.string, "{:?}", value).unwrap();
```

Comment:
Two things here

### Unwrapping
This is an interesting case; according to the documentation, [writing to a string is considered infallible](https://doc.rust-lang.org/std/fmt/struct.Error.html).

There's one case where it very much can fail: we need to allocate memory to hold the string and there's no guarantee there's memory available. However, lots of Rust APIs don't expose this (Vec::with_capacity(), for example, could fail to allocate memory, but it doesn't return a `Result`).

So, this might be a case where it's okay to panic on an error, but I'd recommend using expect to note why it's okay 
```Rust
write!(self.string, "{:?}", value).expect("Writing to a string should never fail");
```

If we're really out of memory, we're likely to be killed by the OOM killer anyway.

### Formatting

This function will get called repeatedly for each field in an event. For example, say there's an event like:
```Rust
tracing::info!(request_id="some-uuid-thingy", count=3, "Some message here");
```

This will be an event with Level=Info and three fields. We'll write the debug representation of each into our string and it'll end up looking something like `Some message heresome-uuid-thingy3` depending on the order fields are processed. You could use something like the example in the tracing documentation:
```Rust
write!(self.string, "{}={:?}; ", field.name(), value).expect("Writing to a string should never fail");
```
which will give you something like `message=Some message here; request_id="some-uuid-thingy"; count=3;`.

We could also emit something like JSON if we want to make it easy to parse by field on the other side, although that adds overhead and we're space-limited here.

Reply:
I've updated the code to use expect() with an explanation to handle the unlikely failure case when writing to a string. I've also adjusted the formatting to include field names in the output, so it now looks like name=value;, which aligns with the tracing documentation example. Thank you for the great suggestion!

JSON formatting would definitely make parsing easier, but I think it’s best to wait until we see how much space a fully-formed tracer setup takes. It may be worth it, or we may find that the name=field format is the best we can do with the space constraints. I'll make a note of this suggestion to address in the future, but for now, I'd lean towards just implementing the formatting change you recommended.